kuma/core/middleware.py

```diff
@@ -1,93 +1,181 @@
+        else:
+            new_path = old_path.replace(script_prefix, new_prefix, 1)
+
+        # Redirect to same path with requested language and without ?lang
+        new_query = dict((smart_str(k), v) for
                          k, v in request.GET.iteritems() if k != 'lang')
+        new_querystring = urlencode(sorted(new_query.items()))
+        new_url = urlunsplit((
+            request.scheme,
+            request.get_host(),
+            new_path,
+            new_querystring,
+            ''  # Fragment / Anchor
+        ))
+        response = HttpResponseRedirect(new_url)
+        add_shared_cache_control(response)
+        return response
 
-            # Never use HttpResponsePermanentRedirect here.
-            # Its a temporary redirect and should return with http 302, not 301
-            response = HttpResponseRedirect(urlparams(new_path, **query))
-            add_shared_cache_control(response)
-            return response
 
-        if full_path != request.path:
-            query_string = request.META.get('QUERY_STRING', '')
-            full_path = urllib.quote(full_path.encode('utf-8'))
+class LocaleStandardizerMiddleware(object):
+    """
+    Convert 404s with legacy locales to redirects.
 
-            if query_string:
-                full_path = '%s?%s' % (full_path, query_string)
+    This should appear higher than LocaleMiddleware in the middleware list.
+    """
+
+    def process_response(self, request, response):
+        """Convert 404s into redirects to language-specific URLs."""
 
-            response = HttpResponseRedirect(full_path)
+        if response.status_code != 404:
+            return response
 
-            # Vary on Accept-Language if we changed the locale
-            old_locale = prefixer.locale
-            new_locale, _ = split_path(full_path)
-            if old_locale != new_locale:
-                response['Vary'] = 'Accept-Language'
+        language_from_path = get_language_from_path(request.path_info)
+        if not language_from_path:
+            # 404 URLs without locale prefixes should remain 404s
+            return response
 
-            add_shared_cache_control(response)
+        literal_from_path = request.path_info.split('/')[1]
+        fixed_locale = None
+        match = literal_from_path == language_from_path
+        lower_match = literal_from_path.lower() == language_from_path.lower()
+        if lower_match and (language_from_path != literal_from_path):
+            # Language code is a lower-case match for a known locale
+            fixed_locale = language_from_path
+        elif literal_from_path.lower() in settings.LOCALE_ALIASES:
+            # Language code is a known general -> specific locale
+            fixed_locale = settings.LOCALE_ALIASES[literal_from_path.lower()]
+        elif not match and literal_from_path.startswith(language_from_path):
+            # Language code is a specific locale (fr vs fr-FR)
+            fixed_locale = language_from_path
```

Comment:
👍 Nice, I verified that `kuma/core/tests/test_locale_middleware.py` covers all three of these cases.